File: backend/app/core/upgrades/stat_modifiers/bullet_speed_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletSpeed1(UpgradeImplementation):
    id = 'bullet_speed_1'
    name = 'Velocity Boost'
    description = '+5% bullet speed'
    rarity = 'common'
    type = 'stat_modifier'
    cost = 2

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.05
        appliers.modifiers.add_multiplier('bullet', 'speed', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/bullet_speed_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const BulletSpeed1Def: UpgradeDef = {
  id: "bullet_speed_1",
  name: "Velocity Boost",
  description: "+5% bullet speed",
  rarity: RarityID.Common,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 2,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Speed,
  value: 0.05,
  isMultiplier: true,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 5,
}

export class BulletSpeed1 extends Upgrade {}
